apps/gyrokinetic.c

```diff
@@ -279,6 +284,52 @@ gkyl_gyrokinetic_app_new(struct gkyl_gk *gk)
       break;
     case GKYL_MIRROR:
       gk_geom_3d = gkyl_gk_geometry_mirror_new(&geometry_inp);
+      if(app->cdim < 3) {
+        gkyl_gk_geometry_c2fa_deflate(gk_geom_3d, &geometry_inp);
+      }
+      double nonuniform_frac = gk->geometry.nonuniform_mapping_fraction;
+      if (nonuniform_frac > 0.0 & nonuniform_frac <= 1.0) {
+        // Copy deflate geometry if necessary
+        if(app->cdim < 3)
+          app->gk_geom = gkyl_gk_geometry_deflate(gk_geom_3d, &geometry_inp);
+        else
+          app->gk_geom = gkyl_gk_geometry_acquire(gk_geom_3d);
+        struct gkyl_array *bmag_global = gkyl_array_new(GKYL_DOUBLE, app->confBasis.num_basis, app->global_ext.volume);
+        if (app->use_gpu) { // Allgather is only a GPU operation, so we must copy these arrays to GPU, then back to CPU
+          struct gkyl_array *bmag_global_dev = mkarr(app->use_gpu, app->confBasis.num_basis, app->global_ext.volume);
+          struct gkyl_array *bmag_dev = mkarr(app->use_gpu, app->confBasis.num_basis, app->local_ext.volume);
+          gkyl_array_copy(bmag_dev, app->gk_geom->bmag);
+          gkyl_comm_array_allgather(app->comm, &app->local, &app->global, bmag_dev, bmag_global_dev);
+          gkyl_array_copy(bmag_global, bmag_global_dev);
+          gkyl_array_release(bmag_global_dev);
+          gkyl_array_release(bmag_dev);
+        }
+        else {
+          gkyl_comm_array_allgather(app->comm, &app->local, &app->global, app->gk_geom->bmag, bmag_global);
+        }
+        geometry_inp.nonuniform_geom = true;
+        geometry_inp.bmag_global = bmag_global;
+        geometry_inp.decomp_basis = app->confBasis;
+        geometry_inp.decomp_grid = app->grid;
+        geometry_inp.decomp_local = app->local;
+        geometry_inp.decomp_local_ext = app->local_ext;
+        geometry_inp.decomp_global = app->global;
+        geometry_inp.decomp_global_ext = app->global_ext;
+        struct gkyl_mirror_geo_c2fa_ctx *c2fa_app = geometry_inp.mirror_geo_c2fa_ctx;
+        gkyl_gk_geometry_release(gk_geom_3d); // release temporary 3d geometry
+        gkyl_gk_geometry_release(app->gk_geom); // release 3d geometry
+        gkyl_array_release(c2fa_app->c2fa);
+        gkyl_array_release(c2fa_app->c2fa_deflate);
+        gk_geom_3d = gkyl_gk_geometry_mirror_new(&geometry_inp);
+        if(app->cdim < 3) {
+          gkyl_gk_geometry_c2fa_deflate(gk_geom_3d, &geometry_inp);
+        }
+        gkyl_array_release(bmag_global);
+        // I don't think I'm releasing the uniform deflated geometry correctly
```

Comment:
This code sequence seems somewhat convoluted. We have multiple deflations of the geometry, quantities being released then reallocated and there is still a final deflation call following. 

At minimum this section of the code needs more comments for why it needs to do things in this order. I understand that there is some complication here because we need to get the magnetic field before we can know what the non-uniform mapping is, so in effect we have to compute the geometry multiple times (once uniform and once non-uniform) but if anything it feels like some of this sequence should not be app-side, some of the sequence should be within the mirror geometry initialization (in the same way that we have a very particular call sequence for initializing geometry on GPUs where things are initialized host-side, then copied over to the device, in effect creating a situation where geometry on GPUs is always from file).